pkg/k8s/watchers/pod.go

```diff
@@ -503,6 +503,9 @@ func (k *K8sWatcher) genServiceMappings(pod *slim_corev1.Pod, podIPs []string, l
 		svcs       []loadbalancer.SVC
 		containers []slim_corev1.Container
 	)
+	if !k8sUtils.IsPodRunning(pod.Status) {
+		return svcs
+	}
```

Comment:
This logic is common to add/update/delete watcher callbacks, so this will leak an entry (<host-ip, host-port>) on a cluster. See http://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L271-L271.

@aanm Why do we need to [call](http://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L220-L220) `deleteK8sPodV1` for non-running pods from the **add** watcher callback? We need to clean up previously saved state only when there is an update for pod state, no?

<img width="421" alt="Screen Shot 2022-12-06 at 4 03 16 PM" src="https://user-images.githubusercontent.com/20311845/206052413-d3de500a-b071-4323-996f-1133703c83c8.png">

Reply:
I see your point about the leaks. I think the alternative would be to add a guard clause everywhere DeleteService is called [here](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L665) and [here](https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L722), to check the backend for the service map entry, and validate that it does in fact equal the ip for the completed pod before deleting. That way we won't break hostPort connectivity for the corresponding `running` pod on the same node. The question is whether there is an easy way to get the service's map entries in `pkg/k8s/watchers/pod.go`? Any thoughts?

Reply:
>The problem essentially boils down to distinguishing between hostPort pods in completed status that have a corresponding running status pod, versus completed pods that don't have a running.

Given that k8s allows only 1 unique <host ip, host port> pod to be deployed at a time, don't we expect that the cilium agent will first receive an event for the completed pod followed by an event for the running pod?

Reply:
> > The problem essentially boils down to distinguishing between hostPort pods in completed status that have a corresponding running status pod, versus completed pods that don't have a running.
> 
> Given that k8s allows only 1 unique <host ip, host port> pod to be deployed at a time, don't we expect that the cilium agent will first receive an event for the completed pod followed by an event for the running pod?

@aditighag no, they can be random.